### PR TITLE
[motherless] Add gallery support

### PR DESCRIPTION
**Pull request by Ti4eeT4e**
### Before submitting a *pull request* make sure you have:
- [x] At least skimmed through [contributing guidelines](https://github.com/yt-dlp/yt-dlp/blob/master/CONTRIBUTING.md#developer-instructions) including [yt-dlp coding conventions](https://github.com/yt-dlp/yt-dlp/blob/master/CONTRIBUTING.md#yt-dlp-coding-conventions)
- [x] [Searched](https://github.com/yt-dlp/yt-dlp/search?q=is%3Apr&type=Issues) the bugtracker for similar pull requests
- [x] Checked the code with [flake8](https://pypi.python.org/pypi/flake8) and [ran relevant tests](https://github.com/yt-dlp/yt-dlp/blob/master/CONTRIBUTING.md#developer-instructions)

### In order to be accepted and merged into yt-dlp each piece of code must be in public domain or released under [Unlicense](http://unlicense.org/). Check one of the following options:
- [x] I am the original author of this code and I am willing to release it under [Unlicense](http://unlicense.org/)
- [ ] I am not the original author of this code but it is in public domain or released under [Unlicense](http://unlicense.org/) (provide reliable evidence)

### What is the purpose of your *pull request*?
- [x] Fix or improvement to an extractor (Make sure to add/update tests)
- [x] New extractor ([Piracy websites will not be accepted](https://github.com/yt-dlp/yt-dlp/blob/master/CONTRIBUTING.md#is-the-website-primarily-used-for-piracy))
- [ ] Core bug fix/improvement
- [ ] New feature (It is strongly [recommended to open an issue first](https://github.com/yt-dlp/yt-dlp/blob/master/CONTRIBUTING.md#adding-new-feature-or-making-overarching-changes))

---

### Description of your *pull request* and other information

Fixes the pattern that the MotherlessIE uses for IDs. The site uses hexadecimal ids, so we can restrict the match here appropriately. Implements supports for galleries, which uses IDs prefixed with `G` or `GV`, which in the previous step we prevented the generic IE from matching.

### CODE REVIEW

File: yt_dlp/extractor/motherless.py [hunk 1]
```diff
@@ -246,3 +246,74 @@ def _get_page(idx):
...
+                video_title=title))
+
+        return entries
+
+    def _real_extract(self, url):
+        gallery_id = self._match_id(url)
+        page_url = compat_urlparse.urljoin(url, f'/GV{gallery_id}')
+        webpage = self._download_webpage(page_url, gallery_id)
+        title = self._search_regex(
+            r'<title>([\w\s]+\w)\s+', webpage, 'title', fatal=False)
```
Comment (pukkandan) [inline yt_dlp/extractor/motherless.py:296]: self._html_extract_title

File: yt_dlp/extractor/motherless.py [hunk 2]
```diff
@@ -246,3 +246,74 @@ def _get_page(idx):
...
+
+        def _get_page(idx):
+            if not page_count:
+                return
+            webpage = self._download_webpage(
+                page_url, gallery_id, query={'page': idx + 1},
+                note='Downloading page %d/%d' % (idx + 1, page_count)
+            )
+            for entry in self._extract_entries(webpage, url):
+                yield entry
```
Comment (pukkandan) [inline yt_dlp/extractor/motherless.py:310]: imo, this should just be part of `_extract_entries`

File: yt_dlp/extractor/motherless.py [hunk 3]
```diff
@@ -246,3 +246,74 @@ def _get_page(idx):
...
+        title = self._search_regex(
+            r'<title>([\w\s]+\w)\s+', webpage, 'title', fatal=False)
+        page_count = self._int(self._search_regex(
+            r'(\d+)</a><a[^>]+rel="next">',
+            webpage, 'page_count', default=1), 'page_count')
+        PAGE_SIZE = 60
+
+        def _get_page(idx):
+            if not page_count:
+                return
```
Comment (pukkandan) [inline yt_dlp/extractor/motherless.py:304]: how can this happen?

File: yt_dlp/extractor/motherless.py (hunk 2 above)
Comment (pukkandan) [inline yt_dlp/extractor/motherless.py:310]:
```suggestion
            yield from self._extract_entries(webpage, url)
```

File: yt_dlp/extractor/motherless.py [hunk 4]
```diff
@@ -246,3 +246,74 @@ def _get_page(idx):
...
+            webpage = self._download_webpage(
+                page_url, gallery_id, query={'page': idx + 1},
+                note='Downloading page %d/%d' % (idx + 1, page_count)
+            )
```
Comment (pukkandan) [inline yt_dlp/extractor/motherless.py:308]:
```suggestion
            webpage = self._download_webpage(
                page_url, gallery_id, query={'page': idx + 1},
                note='Downloading page %d/%d' % (idx + 1, page_count))
```

File: yt_dlp/extractor/motherless.py [hunk 5]
```diff
@@ -246,3 +246,74 @@ def _get_page(idx):
...
+            entries.append(self.url_result(
+                video_url, ie=MotherlessIE.ie_key(), video_id=video_id,
+                video_title=title))
+
+        return entries
+
+    def _real_extract(self, url):
+        gallery_id = self._match_id(url)
+        page_url = compat_urlparse.urljoin(url, f'/GV{gallery_id}')
+        webpage = self._download_webpage(page_url, gallery_id)
```
Comment (pukkandan) [inline yt_dlp/extractor/motherless.py:294]: This webpage is being downloaded twice?

File: yt_dlp/extractor/motherless.py [hunk 6]
```diff
@@ -246,3 +246,74 @@ def _get_page(idx):
...
+                yield entry
+
+        playlist = InAdvancePagedList(_get_page, page_count, PAGE_SIZE)
+
+        return {
+            '_type': 'playlist',
+            'id': gallery_id,
+            'title': title,
+            'entries': playlist
+        }
```
Comment (pukkandan) [inline yt_dlp/extractor/motherless.py:319]:
```suggestion
        return self.playlist_result(
            InAdvancePagedList(_get_page, page_count, PAGE_SIZE), gallery_id, title)
```

**Review by pukkandan — CHANGES_REQUESTED** [review]

File: yt_dlp/extractor/motherless.py (hunk 3 above)
Comment (pukkandan) [reply yt_dlp/extractor/motherless.py:304]: ah, I now see how similar the codes for the 2 extractors are. Why not generalize it into a single base IE?

File: yt_dlp/extractor/motherless.py (hunk 5 above)
Comment (Ti4eeT4e) [reply yt_dlp/extractor/motherless.py:294]: To populate the metadata and `page_count`

File: yt_dlp/extractor/motherless.py (hunk 2 above)
Comment (Ti4eeT4e) [reply yt_dlp/extractor/motherless.py:310]: Which bits exactly? This is the pager function that is being passed into `InAdvancedPagedList`.

File: yt_dlp/extractor/motherless.py [hunk 7]
```diff
@@ -246,3 +246,74 @@ def _get_page(idx):
...
+            title = mobj.group('title')
+            entries.append(self.url_result(
+                video_url, ie=MotherlessIE.ie_key(), video_id=video_id,
+                video_title=title))
+
+        return entries
+
+    def _real_extract(self, url):
+        gallery_id = self._match_id(url)
+        page_url = compat_urlparse.urljoin(url, f'/GV{gallery_id}')
```
[the remark this answers is not in the document]
Comment (Ti4eeT4e) [reply yt_dlp/extractor/motherless.py:293]: Because it would shadow the `url` var passed from the argument list, but that is probably not a problem after matching the id.

File: yt_dlp/extractor/motherless.py (hunk 2 above)
Comment (pukkandan) [reply yt_dlp/extractor/motherless.py:310]: The whole thing. You can use `functools.partial` as needed

File: yt_dlp/extractor/motherless.py [hunk 8]
```diff
@@ -246,3 +246,74 @@ def _get_page(idx):
...
+        webpage = self._download_webpage(page_url, gallery_id)
+        title = self._search_regex(
+            r'<title>([\w\s]+\w)\s+', webpage, 'title', fatal=False)
+        page_count = self._int(self._search_regex(
+            r'(\d+)</a><a[^>]+rel="next">',
+            webpage, 'page_count', default=1), 'page_count')
+        PAGE_SIZE = 60
```
[the remark this answers is not in the document]
Comment (Ti4eeT4e) [reply yt_dlp/extractor/motherless.py]: Done

File: yt_dlp/extractor/motherless.py (hunk 7 above)
[the remark this answers is not in the document]
Comment (pukkandan) [reply yt_dlp/extractor/motherless.py]: I meant: why construct `page_url` and not directly use the `url`

**Comment by pukkandan** [comment]
pls do not force-push. Having a messy commit history is fine

File: yt_dlp/extractor/motherless.py (hunk 3 above)
Comment (Ti4eeT4e) [reply yt_dlp/extractor/motherless.py]: I'll look into that.

File: yt_dlp/extractor/motherless.py (hunk 2 above)
Comment (Ti4eeT4e) [reply yt_dlp/extractor/motherless.py]: Do you have an example extractor I can look at for that?

Comment (pukkandan) [reply yt_dlp/extractor/motherless.py]: ping me once other points are addressed, and I'll help with this

Comment (Ti4eeT4e) [reply yt_dlp/extractor/motherless.py]: I think most other points are.

**Comment by Ti4eeT4e** [comment]
Dropped ABC from the PaginatedIE class to fix the tests.

**Comment by Ti4eeT4e** [comment]
Fixed flake8 errors.

File: yt_dlp/extractor/motherless.py [hunk 9]
```diff
@@ -147,102 +146,88 @@ def _real_extract(self, url):
...
 
 
-class MotherlessGroupIE(InfoExtractor):
-    _VALID_URL = r'https?://(?:www\.)?motherless\.com/gv?/(?P<id>[a-z0-9_]+)'
+class MotherlessPaginatedIE(InfoExtractor):
+    _PAGE_SIZE = 60
+
+    def _extract_entries(self, webpage, base):
+        for mobj in re.finditer(r'href=".*(?P<href>\/[A-F0-9]+)"\s+title="(?P<title>[^"]+)', webpage):
+            video_url = compat_urlparse.urljoin(base, mobj.group("href"))
```
Comment (pukkandan) [inline yt_dlp/extractor/motherless.py:154]:
```suggestion
            video_url = compat_urlparse.urljoin(base, mobj.group('href'))
```

File: yt_dlp/extractor/motherless.py [hunk 10]
```diff
@@ -147,102 +146,88 @@ def _real_extract(self, url):
...
+            video_id = MotherlessIE.get_temp_id(video_url)
+
+            if video_id:
+                yield self.url_result(
+                    video_url, ie=MotherlessIE.ie_key(), video_id=video_id,
+                    video_title=mobj.group('title'))
+
+    def _real_extract(self, url):
+        _id = self._match_id(url)
```
Comment (pukkandan) [inline yt_dlp/extractor/motherless.py:163]: Conventionally, we use `item_id`, but 🤷

File: yt_dlp/extractor/motherless.py [hunk 11]
```diff
@@ -147,102 +146,88 @@ def _real_extract(self, url):
...
+        webpage = self._download_webpage(url, _id)
+        title = self._search_regex(r'^([\w\s]+)\s+', self._html_extract_title(webpage), 'title')
```
Comment (pukkandan) [inline yt_dlp/extractor/motherless.py:165]: Why the search_regex?

Comment (Ti4eeT4e) [reply yt_dlp/extractor/motherless.py:165]: Title looks like so:

```
<title>F3C5049 | MOTHERLESS.COM ™</title>
```

I could also

```
title = "|".join(self._html.extract_title(webpage).split('|')[:-1]
```

Comment (Ti4eeT4e) [reply yt_dlp/extractor/motherless.py:165]: Actually also like so:

```html
<title>Movie Scenes - Videos - Hot and sexy scenes from &quot;regular&quot; movies... Beautiful actresses fully | MOTHERLESS.COM ™</title>
```

where `Movie Scenes` would be the title we're looking for, so the regex works.

Comment (pukkandan) [reply yt_dlp/extractor/motherless.py:165]: Why is "Movie Scenes" the title here? The whole thing except the suffix should be the title. So do `strip_or_none(remove_end(..., '| MOTHERLESS.COM '))`

Comment (Ti4eeT4e) [reply yt_dlp/extractor/motherless.py:165]: It is three things:

- Movie scenes (title)
- Video (meaning videos, as opposed to images)
- Hot and sexy scenes .... (description)

But that is handled inconsistently between groups and galleries.

**Comment by Ti4eeT4e** [comment]
Tests succeed, anything else?

File: yt_dlp/extractor/motherless.py [hunk 12]
```diff
@@ -147,102 +146,81 @@ def _real_extract(self, url):
...
-                video_url, ie=MotherlessIE.ie_key(), video_id=video_id,
-                video_title=title))
-        # Alternative fallback
-        if not entries:
-            entries = [
-                self.url_result(
-                    compat_urlparse.urljoin(base, '/' + entry_id),
-                    ie=MotherlessIE.ie_key(), video_id=entry_id)
-                for entry_id in orderedSet(re.findall(
-                    r'data-codename=["\']([A-Z0-9]+)', webpage))]
```
Comment (pukkandan) [inline yt_dlp/extractor/motherless.py:209]: why was this removed

File: yt_dlp/extractor/motherless.py [hunk 13]
```diff
@@ -147,102 +146,81 @@ def _real_extract(self, url):
...
-        return entries
-
-    def _real_extract(self, url):
-        group_id = self._match_id(url)
-        page_url = compat_urlparse.urljoin(url, '/gv/%s' % group_id)
-        webpage = self._download_webpage(page_url, group_id)
-        title = self._search_regex(
-            r'<title>([\w\s]+\w)\s+-', webpage, 'title', fatal=False)
-        description = self._html_search_meta(
-            'description', webpage, fatal=False)
```
Comment (pukkandan) [inline yt_dlp/extractor/motherless.py:219]: same

File: yt_dlp/extractor/motherless.py [hunk 14]
```diff
@@ -147,102 +146,81 @@ def _real_extract(self, url):
...
-        page_count = self._int(self._search_regex(
-            r'(\d+)</(?:a|span)><(?:a|span)[^>]+rel="next">',
-            webpage, 'page_count', default=0), 'page_count')
-        if not page_count:
-            message = self._search_regex(
-                r'class="error-page"[^>]*>\s*<p[^>]*>\s*(?P<error_msg>[^<]+)(?<=\S)\s*',
-                webpage, 'error_msg', default=None) or 'This group has no videos.'
-            self.report_warning(message, group_id)
```
Comment (pukkandan) [inline yt_dlp/extractor/motherless.py:227]: this too

**Comment by bashonly** [comment]
Closed by f2ff0f6f1914b82d4a51681a72cc0828115dcb4a
